### PR TITLE
Add `getparams` and `setparams!!` following `AbstractMCMC` v5.5 and v5.6

### PR DESCRIPTION
**Pull request by sunxd3**
ref: https://github.com/TuringLang/Turing.jl/issues/2367

### CODE REVIEW

**Comment by sunxd3** [comment]
the test is failing because an numerical test (https://github.com/TuringLang/AdvancedMH.jl/actions/runs/11434545096/job/31870916921?pr=103#step:5:353) 
should we relax the `atol`, and is this new?

File: test/runtests.jl [hunk 1]
```diff
@@ -33,6 +34,15 @@ include("util.jl")
...
     LogDensityProblems.dimension(::typeof(density)) = 2
 
+    @testset "getparams/setparams!! (AbstractMCMC interface)" begin
+        test_spl = StaticMH([Normal(0, 1), Normal(0, 1)])
+        t, _ = AbstractMCMC.step(Random.default_rng(), model, test_spl)
+        @test AbstractMCMC.getparams(t) == t.params
+        @test AbstractMCMC.setparams!!(t, AbstractMCMC.getparams(t)) == t
+        t_replaced = AbstractMCMC.setparams!!(t, (μ=1.0, σ=2.0))
+        @test t_replaced.params == (μ=1.0, σ=2.0)
+    end
```
Comment (penelopeysm) [inline test/runtests.jl:44]:
```suggestion
    @testset "getparams/setparams!! (AbstractMCMC interface)" begin
        t1, _ = AbstractMCMC.step(Random.default_rng(), model, StaticMH([Normal(0, 1), Normal(0, 1)]))
        t2, _ = AbstractMCMC.step(Random.default_rng(), model, MALA(x -> MvNormal(x, I)); initial_params=ones(2))
        for t in [t1, t2]
            @test AbstractMCMC.getparams(t) == t.params
            @test AbstractMCMC.setparams!!(t, AbstractMCMC.getparams(t)) == t
            t_replaced = AbstractMCMC.setparams!!(t, (μ=1.0, σ=2.0))
            @test t_replaced.params == (μ=1.0, σ=2.0)
        end
    end
```

I guess we should test `MALA` too. Does this work?

**Review by penelopeysm — COMMENTED** [review]
The PRNG generates different numbers on different Julia versions even if the seed is set the same, so I guess 1234 was just an unlucky choice 😅

**Review by penelopeysm — APPROVED** [review]

**Comment by penelopeysm** [comment]
Oh, one more thing, should it be 0.9.0 rather than 0.8.4? Since we're introducing new functions.

**Comment by torfjelde** [comment]
> Oh, one more thing, should it be 0.9.0 rather than 0.8.4? Since we're introducing new functions.

It's a bit weird, but basically, at least to my understanding, when we haven't reached version 1 yet, we use minor version as what is considered "breaking" and patch version for everything that isn't. Once we have non-zero major version, then we bump minor version when we introduce new (even non-breaking) functionality. Sooo in this case, I say we stick with bumping patch version.

**Review by torfjelde — CHANGES_REQUESTED** [review]
This PR suffers from the same issues I've raised in AdvancedHMC.jl: gradient information will be out of sync.

This means that if you do something like
```julia
AbstractMCMC.step(model, sampler, setparams!!(state, new_params))
```
you'll use the incorrect gradient in the MH step, resulting in the incorrect chain. I've outlined what I think we should do here: https://github.com/TuringLang/AdvancedHMC.jl/pull/378#pullrequestreview-2393847647

File: Project.toml [hunk 2]
```diff
@@ -23,18 +24,19 @@ AdvancedMHMCMCChainsExt = "MCMCChains"
 AdvancedMHStructArraysExt = "StructArrays"
 
 [compat]
-AbstractMCMC = "5"
+AbstractMCMC = "5.6"
+BangBang = "0.3.19, 0.4"
```
Comment (torfjelde) [inline Project.toml:28]: I'm assuming this is left-over from when you were using BangBang.jl? We're no longer using it, right?

File: src/AdvancedMH.jl [hunk 3]
```diff
@@ -2,6 +2,7 @@ module AdvancedMH
 
 # Import the relevant libraries.
 using AbstractMCMC
+using BangBang
```
Comment (torfjelde) [inline src/AdvancedMH.jl:5]:
```suggestion
```

**Review by torfjelde — CHANGES_REQUESTED** [review]
Looks good! Just need to remove the BangBang.jl references.

**Comment by sunxd3** [comment]
good call! now removed, thanks a lot for reviewing this all these times

**Pull request merged**